src/internet_identity/src/delegation.rs

```diff
@@ -45,15 +45,51 @@ pub async fn prepare_delegation(
     });
     update_root_hash();
 
-    state::usage_metrics_mut(|metrics| {
-        metrics.delegation_counter += 1;
-    });
+    delegation_bookkeeping(frontend);
+
     (
         ByteBuf::from(der_encode_canister_sig_key(seed.to_vec())),
         expiration,
     )
 }
 
+/// Update metrics and the list of latest front-end origins.
+fn delegation_bookkeeping(frontend: FrontendHostname) {
+    state::usage_metrics_mut(|metrics| {
+        metrics.delegation_counter += 1;
+    });
+    update_latest_delegation_origins(frontend);
+}
+
+/// Add the current front-end to the list of latest used front-end origins.
+fn update_latest_delegation_origins(frontend: FrontendHostname) {
+    let now = time();
+
+    persistent_state_mut(|persistent_state| {
+        let latest_delegation_origins = persistent_state
+            .latest_delegation_origins
+            .get_or_insert(HashMap::new());
+
+        if let Some(timestamp) = latest_delegation_origins.get_mut(&frontend) {
+            *timestamp = now;
+        } else {
+            latest_delegation_origins.insert(frontend, now);
+        };
+
+        // drop entries older than 30 days
+        latest_delegation_origins.retain(|_, timestamp| now - *timestamp < 30 * DAY_NS);
+
+        // if we still have too many entries, drop the oldest
+        if latest_delegation_origins.len() as u64
+            > persistent_state.max_num_latest_delegation_origins.unwrap()
+        {
+            let mut values: Vec<_> = latest_delegation_origins.clone().into_iter().collect();
+            values.sort_by(|(_, timestamp_1), (_, timestamp_2)| timestamp_1.cmp(timestamp_2));
+            latest_delegation_origins.remove(&values[0].0);
```

Comment:
It seems a bit unfortunate that once we fill up `latest_delegation_origins` this relatively costly operation will be executed on each insertion of a new origin (i..e clone the values, sort them, remove the single oldest one).  I suppose with the current usage filling up is probably not a concern, but maybe in the future we should change the structure to something that would allow for more efficient removal?  (i.e. I believe there is no need for change right now, but maybe we should add a comment for the future?)

Reply:
I agree. Added a comment.